### PR TITLE
NAS-110951 / 21.06-BETA.1 / Fixed ACL Manager for NFSv4 type, Added 'acltype' to dataset advanced form options

### PR DESCRIPTION
**Pull request by RehanY147**
NAS-110951

### CODE REVIEW

**Review by undsoft — CHANGES_REQUESTED** [review]
Doing `zfs set acltype=nfsv4 tank/share` and then opening ACL Manager appears to be working, but I've found additional issues:
- Previously it was possible to create a dataset without explicitly setting up ACL Type. Now user gets a validation error and has to go to Advanced settings.
- Failed to create dataset: 'acltype' must be one of 'off | nfsv4 | posix' when selecting NFS4ACL. Expected?

**Comment by RehanY147** [comment]
> 
> 
> Doing `zfs set acltype=nfsv4 tank/share` and then opening ACL Manager appears to be working, but I've found additional issues:
> 
>     * Previously it was possible to create a dataset without explicitly setting up ACL Type. Now user gets a validation error and has to go to Advanced settings.
> 
>     * Failed to create dataset: 'acltype' must be one of 'off | nfsv4 | posix' when selecting NFS4ACL. Expected?

Not sure about the second part. I picked these values from the `pool.dataset.create` api call's docs. They should be working.

File: src/app/pages/storage/volumes/datasets/dataset-form/dataset-form.component.ts [hunk 1]
```diff
@@ -649,6 +650,20 @@ export class DatasetFormComponent implements FormConfiguration {
...
+          type: 'select',
+          name: 'acltype',
+          placeholder: T('ACL Type'),
+          options: [
+            { label: 'OFF', value: 'OFF' },
+            { label: 'NOACL', value: 'NOACL' },
+            { label: 'NFSV4', value: 'NFSV4' },
+            { label: 'NFS4ACL', value: 'NFS4ACL' },
+            { label: 'POSIX', value: 'POSIX' },
+            { label: 'POSIXACL', value: 'POSIXACL' },
```
[the remark this answers is not in the document]
Comment (anodos325) [reply src/app/pages/storage/volumes/datasets/dataset-form/dataset-form.component.ts]: NFS4ACL and POSIXACL are old names. They were renamed at some point after initial versions of SCALE and we allowed through for backwards compatibility. Obviously if you're getting errors then that needs to be fixed, but users shouldn't be presented with the options in the GUI.

Requiring explicit acltype seems like a behavior change or regression. Probably related to work on inheritable properties.

File: src/app/enums/dataset-acl-type.enum.ts [hunk 2]
```diff
@@ -1,8 +1,8 @@
...
   Off = 'OFF',
   NoAcl = 'NOACL',
-  Nfsv4 = 'NFSV4',
-  Nfs4Acl = 'NFS4ACL',
-  Posix = 'POSIX',
-  PosixAcl = 'POSIXACL',
+  NFSv4 = 'NFSV4',
+  NFS4Acl = 'NFS4ACL',
+  POSIX = 'POSIX',
+  POSIXAcl = 'POSIXACL',
```
Comment (undsoft) [inline src/app/enums/dataset-acl-type.enum.ts:7]: Previous code style was correct.

File: src/app/pages/storage/volumes/datasets/dataset-form/dataset-form.component.ts [hunk 3]
```diff
@@ -655,14 +655,13 @@ export class DatasetFormComponent implements FormConfiguration {
...
           placeholder: T('ACL Type'),
           options: [
-            { label: 'OFF', value: 'OFF' },
-            { label: 'NOACL', value: 'NOACL' },
-            { label: 'NFSV4', value: 'NFSV4' },
-            { label: 'NFS4ACL', value: 'NFS4ACL' },
+            { label: 'Inherit', value: 'INHERIT' },
+            { label: 'Off', value: 'OFF' },
+            { label: 'NFSv4', value: 'NFSV4' },
             { label: 'POSIX', value: 'POSIX' },
```
Comment (undsoft) [inline src/app/pages/storage/volumes/datasets/dataset-form/dataset-form.component.ts:661]: { label: 'POSIX', value: DatasetAclType.Posix }

File: src/app/pages/storage/volumes/datasets/dataset-form/dataset-form.component.ts [hunk 4]
```diff
@@ -649,6 +651,19 @@ export class DatasetFormComponent implements FormConfiguration {
...
+          name: 'acltype',
+          placeholder: T('ACL Type'),
+          options: [
+            { label: T('Inherit'), value: DatasetAclType.Inherit },
+            { label: T('Off'), value: DatasetAclType.Off },
+            { label: T('NFSv4'), value: DatasetAclType.Nfsv4 },
+            { label: T('POSIX'), value: DatasetAclType.Posix },
+          ],
+          required: false,
+          value: 'INHERIT',
```
Comment (denysbutenko) [inline src/app/pages/storage/volumes/datasets/dataset-form/dataset-form.component.ts:665]: It could be `DatasetAclType.Inherit`

**Comment by anodos325** [comment]
Okay, loaded in testing environment. `Dataset Path` in permissions editor is blank. 

Looks like the ACL editor freezes if user selects one of the POSIX ACL templates on the NFSv4 ACL editor. We should probably remove those from the list for now (POSIX_OPEN, POSIX_RESTRICTED) until I alter  `filesystem.default_acl_choices` to optionally take a path (dataset mountpoint in this case) as an argument.

**Comment by anodos325** [comment]
Looks like in my tests when filesystem.getacl returns the following, the corresponding named entry (for gid 545) is not populated in the webui:
```
{"acl": [{"tag": "owner@", "id": -1, "perms": {"BASIC": "FULL_CONTROL"}, "flags": {"BASIC": "INHERIT"}, "type": "ALLOW"}, {"tag": "group@", "id": -1, "perms": {"BASIC": "FULL_CONTROL"}, "flags": {"BASIC": "INHERIT"}, "type": "ALLOW"}, {"tag": "everyone@", "id": -1, "perms": {"BASIC": "MODIFY"}, "flags": {"BASIC": "INHERIT"}, "type": "ALLOW"}, {"tag": "GROUP", "id": 545, "perms": {"BASIC": "MODIFY"}, "flags": {"BASIC": "INHERIT"}, "type": "ALLOW"}], "trivial": false, "uid": 666, "gid": 666, "path": "/mnt/dozer/NFS4", "nfs41_flags": {"protected": false, "autoinherit": false}, "acltype": "NFS4"}
```

**Comment by undsoft** [comment]
Retested again: Path fields are populated, Posix presets are hidden, was able to edit both posix and nfs permissions.
Additional fixes for some inputs in aces going blank is coming in this PR: https://github.com/truenas/webui/pull/5566
Merging this one.

**Pull request merged**